lua/cfc_ulx_commands/curse/cl_utils.lua

```diff
@@ -1,4 +1,85 @@
------ SETUP ----
+CFCUlxCurse = CFCUlxCurse or {}
+CFCUlxCurse._marchFolderCache = CFCUlxCurse._marchFolderCache or {}
+CFCUlxCurse._marchFolderCacheDroppedRoot = CFCUlxCurse._marchFolderCacheDroppedRoot or {}
+
+
+local marchFolderCache = CFCUlxCurse._marchFolderCache
+local marchFolderCacheDroppedRoot = CFCUlxCurse._marchFolderCacheDroppedRoot
+
+local fileFind = file.Find
+local tableInsert = table.insert
+local stringSub = string.sub
+
+
+--[[
+    - Returns a sequential list of all files in a folder, recursively.
+    - Includes the full file path and extension for each file.
+    
+    folder: (string)
+        - The folder to march. Exclude the trailing slash.
+        - sound, models, materials/models, etc.
+    path: (string)
+        - https://wiki.facepunch.com/gmod/File_Search_Paths
+        - GAME, DATA, etc.
+    out: (optional) (table)
+        - The table to append the file paths to.
+        - If nil, a new table will be created.
+
+    RETURNS: out (table)
+        - A table of file paths.
+--]]
+function CFCUlxCurse.MarchFolder( folder, path, out )
+    folder = folder .. "/"
+    out = out or {}
+
+    local fileNames, folderNames = fileFind( folder .. "*", path )
+
+    for _, fileName in ipairs( fileNames ) do
+        tableInsert( out, folder .. fileName )
+    end
+
+    for _, folderName in ipairs( folderNames ) do
+        CFCUlxCurse.MarchFolder( folder .. folderName, path, out )
+    end
+
+    return out
+end
+
+--[[
+    - Similar to MarchFolder, but caches the results for each unique root folder/path combination.
+    - Provide forceRecahce as true to ignore the cache and recalculate the files.
+    - Prodive dropRoot as true to strip 'folder/' from the start of each file path.
+        - This being true/false results in using a different cache.
+--]]
```

Comment:
I think `forceRecache` and `dropRoot` are bad params - the caller should do that on their own, keeping this function's responsibilities minimal (the first principal of [SOLID](https://en.wikipedia.org/wiki/Single-responsibility_principle))
```suggestion
--- Recursively gathers a list of all files in a given directory and caches the result
--- @param folder string The folder to match, excluding the trailing slash (e.g. "sound", "models", etc.)
--- @param path string Search Path (e.g. "GAME", "DATA", etc.)
--- @param forceRecache boolean? Whether or not to skip the cache (default: false)
--- @param dropRoot boolean? Whether or not to drop the first folder from each file path
--- @return string[] out The collected file paths
```